src/main/java/io/stargate/sgv2/jsonapi/service/cqldriver/CqlSessionFactory.java

```diff
@@ -0,0 +1,157 @@
+      DatabaseType databaseType,
+      String localDatacenter,
+      List<String> cassandraEndPoints,
+      Integer cassandraPort,
+      List<SchemaChangeListener> schemaChangeListeners) {
+    this(
+        applicationName,
+        databaseType,
+        localDatacenter,
+        cassandraEndPoints,
+        cassandraPort,
+        schemaChangeListeners,
+        CqlSessionBuilder::new);
+  }
+
+  /**
+   * Constructor for the CqlSessionFactory, normally use the overload without the
+   * sessionBuilderSupplier param.
+   *
+   * @param applicationName the name of the application, set on the CQL session
+   * @param databaseType the type of database, controls contact points and other settings
+   * @param localDatacenter the local datacenter for the client connection.
+   * @param cassandraEndPoints the Cassandra endpoints, only used when the database type is
+   *     CASSANDRA
+   * @param cassandraPort the Cassandra port, only used when the database type is CASSANDRA
+   * @param schemaChangeListeners the schema change listeners, these are added to the session to
+   *     listen for schema changes from it.
+   * @param sessionBuilderSupplier a supplier for creating CqlSessionBuilder instances, so that
+   *     testing can mock the builder for session creation. In prod code use the ctor without this.
+   */
+  CqlSessionFactory(
+      String applicationName,
+      DatabaseType databaseType,
+      String localDatacenter,
+      List<String> cassandraEndPoints,
+      Integer cassandraPort,
+      List<SchemaChangeListener> schemaChangeListeners,
+      Supplier<CqlSessionBuilder> sessionBuilderSupplier) {
+
+    this.applicationName =
+        Objects.requireNonNull(applicationName, "applicationName must not be null");
+    if (applicationName.isBlank()) {
+      throw new IllegalArgumentException("applicationName must not be blank");
+    }
+    this.databaseType = Objects.requireNonNull(databaseType, "databaseType must not be null");
+    this.localDatacenter =
+        Objects.requireNonNull(localDatacenter, "localDatacenter must not be null");
+
+    this.schemaChangeListeners =
+        schemaChangeListeners == null ? List.of() : List.copyOf(schemaChangeListeners);
+    this.sessionBuilderSupplier =
+        Objects.requireNonNull(sessionBuilderSupplier, "sessionBuilderSupplier must not be null");
+
+    // these never change, and we do not have them in astra, so we can cache
+    if (databaseType == DatabaseType.CASSANDRA) {
+      Objects.requireNonNull(cassandraEndPoints, "cassandraEndPoints must not be null");
+      if (cassandraEndPoints.isEmpty()) {
+        throw new IllegalArgumentException(
+            "Database type is %s but cassandraEndPoints is empty.".formatted(databaseType));
+      }
+      contactPoints =
+          cassandraEndPoints.stream()
+              .map(host -> new InetSocketAddress(host, cassandraPort))
+              .toList();
+    } else {
+      contactPoints = List.of();
```

Comment:
Should this actually verify caller does NOT pass non-null/non-empty `cassandraEndPoints` -- given it is ignored completely for Astra/offline cases?